controller/api/destination/external_endpoint_manager.go

```diff
@@ -0,0 +1,275 @@
+	em.lec = leaderelection.LeaderElectionConfig{
+		// When runtime context is cancelled, lock will be released. Implies any
+		// code guarded by the lease _must_ finish before cancelling.
+		ReleaseOnCancel: true,
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      leaseName,
+				Namespace: controllerNs,
+			},
+			Client: k8sAPI.Client.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: hostname,
+			},
+		},
+		LeaseDuration: leaseDuration,
+		RenewDeadline: leaseRenewDeadline,
+		RetryPeriod:   leaseRetryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(context.Context) {
+				em.Lock()
+				defer em.Unlock()
+				em.isLeader = true
+			},
+			OnStoppedLeading: func() {
+				em.Lock()
+				defer em.Unlock()
+				em.isLeader = false
+				em.log.Infof("%s released lease", hostname)
+			},
+			OnNewLeader: func(identity string) {
+				if identity == hostname {
+					em.log.Infof("%s acquired lease", hostname)
+				}
+			},
+		},
+	}
+
+	return em, nil
+}
+
+// Start will run the endpoint manager's processing loop and leader elector.
+//
+// The function will spawn two background tasks; one to run the leader elector
+// client that and one that will process updates applied by the informer
+// callbacks.
+func (em *EndpointManager) Start() {
+	// Create a parent context that will be used by leader elector to gracefully
+	// shutdown.
+	//
+	// When cancelled (either through cancel function or by having its done
+	// channel closed), the leader elector will release the lease and stop its
+	// execution.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		for {
+			// Block until a lease is acquired or a lease has been released
+			leaderelection.RunOrDie(ctx, em.lec)
+			// If the context has been cancelled, exit the function, otherwise
+			// continue spinning.
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+		}
+	}()
+
+	// Start a background task to process updates. When a shutdown signal is
+	// received over the manager's stop channel, it is propagated to the elector
+	// through the context object to ensure the elector task does not leak.
+	go func() {
+		for {
+			select {
+			case update := <-em.updates:
+				em.processUpdate(update)
+			case <-em.stop:
+				em.log.Info("Received shutdown signal")
+				// Propagate shutdown to elector through the context.
+				cancel()
+				return
+			}
+		}
+	}()
+}
+
+// enqueueUpdate will enqueue a service key to the updates buffer to be
+// processed by the endpoint manager. When a write to the buffer blocks, the
+// update is dropped.
+func (em *EndpointManager) enqueueUpdate(update string) {
+	select {
+	case em.updates <- update:
+		// Update successfully enqueued.
+	default:
+		// Drop update
+		// TODO (matei): add overflow counter
+		em.log.Debugf("External endpoint manager queue is full; dropping update for %s", update)
+		return
```

Comment:
You mentioned that you have concerns about simply dropping the update on the floor. Let's think about that (I might be totally wrong). The whole machinery that is present in the destinations service is there because the sink of our updates is a gRPC stream that gets into a state where there is a pending write that blocks the whole pipeline and that **cannot be canceled and will not timeout.**. This last bit is the main reason why the queue there might backup. In my mind, we can avoid getting into a similar state here by: 

- ensuring there is a timeout on the k8s api client that will fail the update operation in case it takes too long
- ensuring the size of the queue is large enough to accommodate a service update (we need to do some back-of-the-envelope math to figure that one out).

In addition to that we need to think whether we can prune these updates. What do we care about in a Service in our case? Do we care if annotations change? Probably not. So thinking through what is a real update vs a resync behaviour.

Does that help?